### ISSUE TITLE
Recommended way of dealing with large amounts of text in a scrollable viewport?

### ISSUE THREAD

**Issue by saurabhnanda** [jtdaugherty/brick#280]
I'm trying to build a scrollable viewport in my Brick application which will "stream" the STDOUT of a shell process running in a separate thread. 

What's the recommended efficient way to get this done? Here's what I'm worried about:

- Should I use `txtWrap` or `strWrap` as the underlying Widget?
- Each time I read the STDOUT handle, I'll get a `String`. Should I keep appending to the underlying `Text` or `String` in the application state? Won't that cause too much allocation?
- Is there any way to write a Widget that uses a bytestring-builder or a lazy-bytestring instead of Text/String?

**Comment by jtdaugherty** [comment]
Hi @saurabhnanda, to do something like this I recommend avoiding the `viewport` primitive entirely and writing a [custom widget](https://github.com/jtdaugherty/brick/blob/master/docs/guide.rst#implementing-custom-widgets) that displays the lines of text based on the size of the drawing region. The `viewport` primitive is only going to work well when the contents can be [render-cached](https://github.com/jtdaugherty/brick/blob/master/docs/guide.rst#the-rendering-cache) and aren't too big; in your case that would depend on whether you want to do line-wrapping (in which case the cache will need to be invalidated on each resize).

When I've implemented this sort of thing before, the way I usually do it is:
* Maintain the content I want to display as a sequence of elements, usually lines of text.
* At render time, render from the end of the sequence backward until I've rendered enough to use up the available space (according to the [rendering context](https://github.com/jtdaugherty/brick/blob/master/docs/guide.rst#using-the-rendering-context)).
* If scrolling backwards to older content is needed, that gets tricky to do and usually the best I've been able to do is to just maintain a line offset in my application state and use that to decide which line to start rendering at.

> Should I use txtWrap or strWrap as the underlying Widget?

You could, although that would depend on whether you are okay with not preserving the `stdout` contents exactly as they originally appeared. Whether you do line-wrapping will also impact how you do rendering cache management, like I mentioned above.

> Each time I read the STDOUT handle, I'll get a String. Should I keep appending to the underlying Text or String in the application state? Won't that cause too much allocation?

Since implementing a scrolling view of the output lines would require splitting the output on line endings anyway, I would suggest an efficient sequence type that supports constant-type append or prepend, such as the one provided by `Data.Sequence`. As for allocation, if you are looking to buffer the output of some application then you are definitely signing up for some allocation just by wanting to keep all the `stdout` output around. You could limit that by keeping around the most recent N lines of output, for example.

> Is there any way to write a Widget that uses a bytestring-builder or a lazy-bytestring instead of Text/String?

You could, but ultimately you will need to convert the bytes to some Vty-representable form, namely `String` or `Text`, because it is not really legal to output arbitrary bytes to the terminal while it is under Vty's control since those bytes could get interpreted by the terminal and thus cause rendering corruption and break Vty's rendering computations. This is less of a Vty/Brick issue and a more general property of terminal output that you'd need to deal with in any terminal application.

**Comment by saurabhnanda** [comment]
> You could, but ultimately you will need to convert the bytes to some Vty-representable form, namely `String` or `Text`

I was investigating this last night, because I was curious why `brick` didn't have a bytestringWrap, or lazyTextWrap. IIUC, the underlying building block is a VTY `Image`, right?

There seem to be many ways to construct an `Image`, for example:

- [`Graphics.Vty.Image.text`](https://hackage.haskell.org/package/vty-5.29/docs/Graphics-Vty-Image.html#v:text) to create an `Image` from a lazy-text.
- [`Graphics.Vty.Image.utf8Bytestring`](https://hackage.haskell.org/package/vty-5.29/docs/Graphics-Vty-Image.html#v:utf8Bytestring) to create an `Image` from a lazy-bytestring
- there are a few more

Upon further inspection it seems that Vty internally converts all of these into **lazy text**, due to the following:

```haskell
type DisplayText = TL.Text

data Image = HorizText { attr :: Attr, displayText :: DisplayText, outputWidth :: Int, charWidth :: Int }
           | ...
```

So, a related question is, why doesn't Brick expose a way to construct `Widget`s using lazy-text (or did I miss it)?

**Comment by saurabhnanda** [comment]
> Since implementing a scrolling view of the output lines would require splitting the output on line endings anyway, I would suggest an efficient sequence type that supports constant-type append or prepend, such as the one provided by Data.Sequence

How would you manage rendering _up to_ `H` lines of text, `W` chars wide, where lines longer than `W` chars are wrapped?

**Comment by jtdaugherty** [comment]
> I was investigating this last night, because I was curious why brick didn't have a bytestringWrap, or lazyTextWrap. IIUC, the underlying building block is a VTY Image, right?

That's right; under the hood, `brick` builds Vty images. As for lack of bytestring functions, that's because they're not the right abstraction for terminal output in Vty.

> So, a related question is, why doesn't Brick expose a way to construct Widgets using lazy-text (or did I miss it)?

Mostly because I didn't think of it, but if you need to embed lazy text values in the mean time, `brick` provides [a function to embed arbitrary Vty images](http://hackage.haskell.org/package/brick-0.55/docs/Brick-Widgets-Core.html#v:raw).

> How would you manage rendering up to H lines of text, W chars wide, where lines longer than W chars are wrapped?

Assuming you want to render the lines with the first lines visible (i.e. truncate at the end of the sequence), then:

Given a list of lines `ls :: [Text]`, a vertical line limit `v`, and a width limit `w`, this can be accomplished with `hLimit h $ vLimit v $ vBox $ txtWrap <$> ls`.

**Comment by saurabhnanda** [comment]
> Assuming you want to render the lines with the first lines visible (i.e. truncate at the end of the sequence), then: Given a list of lines ls :: [Text], a vertical line limit v, and a width limit w, this can be accomplished with hLimit h $ vLimit v $ vBox $ txtWrap <$> ls.

Allow me to rephrase my question. What's the recommended data structure for implementing essentially a `tail` with a scroll-back buffer of a few thousand lines, at least, if not more.

Should one maintain a list of texts (as you've suggested), keep cons-ing new output on top of it (for efficient appends), take the first H elements, reverse them, and use them to construct a `vBox` inside a viewport? How much allocation is this going to do for every render? How expensive will it be to implement a scroll-back feature?

**Comment by jtdaugherty** [comment]
> Should one maintain a list of texts (as you've suggested), keep cons-ing new output on top of it (for efficient appends)

Yes, I think this would work well. If you are concerned about appending to either end, `Data.Sequence` is a good choice.

> take the first H elements, reverse them, and use them to construct a vBox inside a viewport?

If you are consing onto the end, you won't need to reverse; just take elements off the end and draw them bottom-up until you have filled the window. As for viewports, no, a viewport would not be needed and you'd need to write a custom widget like I mentioned above in order to get access to the rendering context. I'll see if I can spend some time writing some helper code for this.

> How much allocation is this going to do for every render?

What is your concern here? (As Haskell is a garbage-collected language we do not often worry about the amount of allocation unless it is egregious.)

> How expensive will it be to implement a scroll-back feature?

Can you say a bit more about what kind of expense you want to measure?

**Comment by saurabhnanda** [comment]
Okay - I looked at `Data.Sequence` at it covers two important concerns that I had -- (a) efficiently consing to the end of the data-structure, and (b) efficiently dropping element from the front of the data-structure (as the buffer gets full).

So, I'll summarize the discussion to ensure that I haven't misunderstood anything:

- maintain an internal buffer in `Data.Sequence String`
- write a custom widget so that you have access to the `context`
- based on `context ^. availHeightL` take `N` strings from end of the `Data.Sequence` and convert them into a `vBox` of string widgets.

this leads to the final question related to:

> At render time, render from the end of the sequence backward until I've rendered enough to use up the available space (according to the rendering context).

Can one call `getContext` multiple times inside a `mapM` loop? Will it give back a different value of `context ^. availHeightL` each time?

Finally, is this a common problem (i.e. tailing the STDOUT of a process and displaying it in a brick UI), to justify an in-built widget within the library itself?

**Comment by jtdaugherty** [comment]
> based on context ^. availHeightL take N strings from end of the Data.Sequence and convert them into a vBox of string widgets.

That's right, provided that you don't want to do line wrapping. Then you'll have to render each string with wrapping and determine how many rows are left, stopping when you run out of room.

> Can one call getContext multiple times inside a mapM loop? Will it give back a different value of context ^. availHeightL each time?

The context maintains the same value each time you call it from within a rendering function; you'll have to accumulate your own rendering results until the sum of their heights reaches `availHeightL`.

> Finally, is this a common problem (i.e. tailing the STDOUT of a process and displaying it in a brick UI), to justify an in-built widget within the library itself?

Yes... and no. On one hand, it is somewhat common for people to run into the problem of how to render a `tail`-style list as you are wanting to do here. (A more common setting is in rendering chat messages.) But on the other hand, the details vary quite a bit. I am definitely thinking about adding something to the library to help with this, but it would provide only this functionality (not allowing scrolling up, for example). I haven't yet decided on an API, but it might look something like
```haskell
f :: (t a -> Maybe (a, t a)) -> (a -> Widget n) -> t a -> Widget n
```
(Arguments would be: a safe `head` function, a widget constructor, and a sequence of elements.)

**Comment by jtdaugherty** [comment]
@saurabhnanda Any thoughts on the demo program I added? (It is also now released in `0.56`.)

**Comment by jtdaugherty** [comment]
@saurabhnanda I'm going to close this due to inactivity, but please re-open if you'd like to continue the discussion. In the mean time, please take a look at `programs/TailDemo.hs`. Thanks!
